### PR TITLE
Add more customisation on helm-chart

### PR DESCRIPTION
**Pull request by ThomasSamson**
Hello, 

We have deployed iTOP on our kubernetes cluster and here is the modifications we did. 
There is, more customisation and default apache port customisation possibility. 

I did a little change on the dockerfile to include port customisation (not in this PR, if you want it I'll push it)

### CODE REVIEW

File: helm-chart/values.yaml [hunk 1]
```diff
@@ -7,7 +7,7 @@ replicaCount: 1
 image:
   repository: supervisions/itop
   pullPolicy: Always
-  tag: 3.0
+  tag: 3.0.2
```
Comment (Hipska) [inline helm-chart/values.yaml]: This is effectively the same. The current version of tag 3.0 is always the latest version 3.0.x

File: helm-chart/templates/deployment.yaml [hunk 2]
```diff
@@ -25,10 +27,10 @@ spec:
...
       {{- end }}
       serviceAccountName: {{ include "itop-chart.serviceAccountName" . }}
-      securityContext:
-        runAsUser: 1000
-        runAsGroup: 3000
-        fsGroup: 2000
+#      securityContext:
+#        runAsUser: 1000
+#        runAsGroup: 3000
+#        fsGroup: 2000
```
Comment (Hipska) [inline helm-chart/templates/deployment.yaml]: Do not comment lines, better to remove them. Git history will keep them for you.

File: helm-chart/values.yaml [hunk 3]
```diff
@@ -41,9 +43,21 @@ securityContext: {}
   # runAsNonRoot: true
   # runAsUser: 1000
 
+pvc:
+  storageClassName: standard
+  conf:
+    size: 1Gi
+  data:
+    size: 1Gi
```
Comment (Hipska) [inline helm-chart/values.yaml]: This will be a little bit small for the data directory as it will contain backups, cache, exports...

File: helm-chart/values.yaml [hunk 4]
```diff
@@ -41,9 +43,21 @@ securityContext: {}
...
+  env:
+    size: 1Gi
+  log:
+    size: 1Gi
```
Comment (Hipska) [inline helm-chart/values.yaml:56]: Same here, this might be too small

File: helm-chart/templates/deployment.yaml [hunk 5]
```diff
@@ -68,18 +70,32 @@ spec:
...
           livenessProbe:
             httpGet:
-              path: /webservices/status.php
+              path: /setup/
               port: http
           readinessProbe:
             httpGet:
-              path: /webservices/status.php
+              path: /setup/
               port: http
```
Comment (Hipska) [inline helm-chart/templates/deployment.yaml]: Why this change? The status should report if iTop is healthy.

File: helm-chart/templates/deployment.yaml [hunk 6]
```diff
@@ -68,18 +70,32 @@ spec:
...
+        volumeMounts:
+        - mountPath: "/var/www/html/data"
+          name: itop-data-pv
+        - mountPath: "/var/www/html/env-production"
+          name: itop-env-pv
+        - mountPath: "/var/www/html/log"
+          name: itop-log-pv
+        - mountPath: "/var/www/html/conf"
+          name: itop-conf-pv
+
```
Comment (Hipska) [inline helm-chart/templates/deployment.yaml]: Why is this needed? The directories already should have this owner and they are also already defined as volume by the Dockerfile.

File: helm-chart/templates/service.yaml [hunk 7]
```diff
@@ -8,7 +8,7 @@ spec:
   type: {{ .Values.service.type }}
   ports:
     - port: {{ .Values.service.port }}
-      targetPort: http
+      targetPort: {{ .Values.service.targetPort }}
```
Comment (Hipska) [inline helm-chart/templates/service.yaml:11]: `.Values.service.targetPort` is defined as `80`. Was the previous value `http` a typo?

**Review by Hipska — CHANGES_REQUESTED** [review]

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: These directory was had root owner by default

File: helm-chart/templates/service.yaml (hunk 7 above)
Comment (ThomasSamson) [reply helm-chart/templates/service.yaml:11]: 80 or http are same values. It's just the default values, on my side I overwrite this with 8080

File: helm-chart/templates/deployment.yaml (hunk 5 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: I don't know why but this page sent 500 error.

File: helm-chart/values.yaml (hunk 4 above)
Comment (ThomasSamson) [reply helm-chart/values.yaml:56]: It's just a default value, but I can increase it if you want :)
Don't see the ITOP recommandation

File: helm-chart/values.yaml (hunk 3 above)
Comment (ThomasSamson) [reply helm-chart/values.yaml]: It's just a default value, but I can increase it if you want :)
Don't see the ITOP recommandation

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: That must be some problem by your setup then:
```
$ docker run --rm supervisions/itop:3.0.2 ls -l
total 1128
drwxr-xr-x  3 www-data www-data    4096 Sep 14 08:29 addons
drwxr-xr-x  4 www-data www-data    4096 Sep 14 08:29 application
-rw-rw-r--  1 www-data www-data     952 Sep 12 14:08 approot.inc.php
-rw-rw-r--  1 www-data www-data    3308 Dec 21  2021 bootstrap.inc.php
drwxr-xr-x  3 www-data www-data    4096 Oct 20 14:55 conf
drwxr-xr-x  3 www-data www-data    4096 Sep 14 08:29 core
drwxr-xr-x  8 www-data www-data    4096 Sep 14 08:29 css
drwxr-xr-x  2 www-data www-data    4096 Oct 20 14:55 data
drwxr-xr-x  3 www-data www-data    4096 Sep 14 08:29 datamodels
drwxr-xr-x  3 www-data www-data    4096 Sep 14 08:29 dictionaries
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 env-production
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 extensions
drwxr-xr-x  9 www-data www-data    4096 Sep 14 08:29 images
-rw-rw-r--  1 www-data www-data    1368 Dec 21  2021 index.php
drwxr-xr-x 12 www-data www-data    4096 Sep 14 08:29 js
drwxr-xr-x 24 www-data www-data    4096 Sep 14 08:29 lib
drwxr-xr-x  2 www-data www-data    4096 Oct 20 14:55 log
-rw-rw-r--  1 www-data www-data 1046438 Sep 12 14:09 manifest.xml
drwxr-xr-x 14 www-data www-data    4096 Sep 14 08:29 node_modules
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 pages
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 portal
drwxr-xr-x  6 www-data www-data    4096 Sep 14 08:29 setup
drwxr-xr-x  8 www-data www-data    4096 Sep 14 08:29 sources
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 synchro
drwxr-xr-x  5 www-data www-data    4096 Sep 14 08:29 templates
-rw-rw-r--  1 www-data www-data     615 Dec 21  2021 web.config
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 webservices
```

File: helm-chart/templates/deployment.yaml (hunk 5 above)
Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: Maybe better to find that out.

File: helm-chart/values.yaml (hunk 3 above)
Comment (Hipska) [reply helm-chart/values.yaml]: Given that even for smallest installations [they suggest](https://www.itophub.io/wiki/page?id=3_0_0:install:requirements) to have 10G for MySQL and this directory volume is the place where backups of that are stored, I would also pick 10 as minimum default value.

File: helm-chart/values.yaml (hunk 4 above)
Comment (Hipska) [reply helm-chart/values.yaml:56]: Okay, I have no experience with helm, but what if a user starts their iTop with this default and the logs become this size? Will that be blocking (disk full)? Will they be able to resize easily?

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: By default kubernetes mount volume as root

File: helm-chart/values.yaml (hunk 4 above)
Comment (ThomasSamson) [reply helm-chart/values.yaml:56]: Yes, when you want deploy itop with helm chart you'll have to set some option, ingress host addr, bdd creds etc .. 
And now you can also overwrite all values present in values.yaml. 

And if you install with 1Gi volumes, you can at any moment upgrade size of volume (please see kube [doc](https://kubernetes.io/docs/concepts/storage/storage-classes/#allow-volume-expansion)

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: Hmm, I did a quick search and found that is related to the `securityContext` which you have removed in this PR.

**Comment by ThomasSamson** [comment]
If its more simple fore you, reject the PR and focus your time on the docker images. 
On our hand we will publish new itop helm chart based on your Docker image. As you like :)

File: helm-chart/values.yaml (hunk 4 above)
Comment (Hipska) [reply helm-chart/values.yaml:56]: Okay thanks, its is fine for me in that case.

**Comment by Hipska** [comment]
No it is fine for me to update the helm chart, I just want to be sure about some choices as I'm not familiar with Helm / Kubernetes.

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: www-data have the userid 33 not 1000.

File: helm-chart/values.yaml (hunk 3 above)
Comment (ThomasSamson) [reply helm-chart/values.yaml]: Size increased :)

File: helm-chart/values.yaml [hunk 8]
```diff
@@ -46,9 +46,9 @@ securityContext: {}
 pvc:
   storageClassName: standard
   conf:
-    size: 1Gi
+    size: 10Gi
```
Comment (Hipska) [inline helm-chart/values.yaml]: How will a config file will ever get this big?

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: Okay, so a matching security context needs to be set instead of removing it I assume?

File: helm-chart/values.yaml (hunk 8 above)
Comment (ThomasSamson) [reply helm-chart/values.yaml]: It was not recommandation? https://github.com/Super-Visions/docker-itop/pull/6#discussion_r1000766224

Comment (Hipska) [reply helm-chart/values.yaml]: No, we were talking about the `data` volume. 1Gi for `config` wil be more than enough..

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: I guess, In this case the container running as root and Apache is running as www-data.

File: helm-chart/values.yaml (hunk 8 above)
Comment (ThomasSamson) [reply helm-chart/values.yaml]: My bad, fixed :)

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: Yes, so if I understand well, I think all should be running fine when only `fsGroup` is set to `33`?

File: helm-chart/templates/deployment.yaml (hunk 5 above)
Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: What is the message the page is sending or what do you see in the logs?

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: I have 500, I was searching about documentation to know if that page is available before first setup. 
Because now all is ok with that file. 
`curl localhost:8080/webservices/status.php
{"status":"RUNNING","code":0,"message":""}`

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: I don't know the effect when you set only the fsgroup. 
The init container is really quick because pulling same image and juste running a chown. 
It's a good way to do that kind of action on containers :)

File: helm-chart/templates/deployment.yaml (hunk 5 above)
Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: The page is indeed giving php error when iTop is not yet installed..

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: And these PHP error send wich kind of http error ? 500 ?

Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: Hmm, no it was a `200 OK` 🤔

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: Can you test so we are sure? I would prefer that instead of a workaround.

File: helm-chart/values.yaml [hunk 9]
```diff
@@ -41,9 +43,21 @@ securityContext: {}
...
+  env:
+    size: 1Gi
+  log:
+    size: 1Gi
+
 service:
   type: ClusterIP
   port: 80
+  targetPort: 80
```
Comment (finaldzn) [inline helm-chart/values.yaml:61]: If you add a values for the port why not add it for the service as well ?

{{ .Values.container.port }}

Comment (ThomasSamson) [reply helm-chart/values.yaml:61]: Hello,

I'm not sure to understand, the port is custom, in the `service.yaml` and in `deployment.yaml`.
Did I forget something ?

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: Set `fsgroup`to 33 is not efficient please see : 
```
root@itop-itop-chart-b6c4f947f-jnpp8:/var/www/html# ls -l
total 1128
drwxr-xr-x  3 www-data www-data    4096 Sep 14 08:29 addons
drwxr-xr-x  4 www-data www-data    4096 Sep 14 08:29 application
-rw-rw-r--  1 www-data www-data     952 Sep 12 14:08 approot.inc.php
-rw-rw-r--  1 www-data www-data    3308 Dec 21  2021 bootstrap.inc.php
drwxrwsr-x  3 root     www-data    4096 Oct 24 08:08 conf
drwxr-xr-x  3 www-data www-data    4096 Sep 14 08:29 core
drwxr-xr-x  8 www-data www-data    4096 Sep 14 08:29 css
drwxrwsr-x  3 root     www-data    4096 Oct 24 08:08 data
drwxr-xr-x  3 www-data www-data    4096 Sep 14 08:29 datamodels
drwxr-xr-x  3 www-data www-data    4096 Sep 14 08:29 dictionaries
drwxrwsr-x  3 root     www-data    4096 Oct 24 08:08 env-production
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 extensions
drwxr-xr-x  9 www-data www-data    4096 Sep 14 08:29 images
-rw-rw-r--  1 www-data www-data    1368 Dec 21  2021 index.php
drwxr-xr-x 12 www-data www-data    4096 Sep 14 08:29 js
drwxr-xr-x 24 www-data www-data    4096 Sep 14 08:29 lib
drwxrwsr-x  3 root     www-data    4096 Oct 24 08:08 log
-rw-rw-r--  1 www-data www-data 1046438 Sep 12 14:09 manifest.xml
drwxr-xr-x 14 www-data www-data    4096 Sep 14 08:29 node_modules
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 pages
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 portal
drwxr-xr-x  6 www-data www-data    4096 Sep 14 08:29 setup
drwxr-xr-x  8 www-data www-data    4096 Sep 14 08:29 sources
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 synchro
drwxr-xr-x  5 www-data www-data    4096 Sep 14 08:29 templates
-rw-rw-r--  1 www-data www-data     615 Dec 21  2021 web.config
drwxr-xr-x  2 www-data www-data    4096 Sep 14 08:29 webservices

```

Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: But is that actually a problem? As it seems the group has the same permissions as the owner of this directory?

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: Yes it is, itop is running as www-data and then the application can't write conf in these files

Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: I don't see the problem as user `www-data` is in the group `www-data` and that group can read and write to the log dir..

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: Yes I'm just seeing this, let me do more test :)

File: helm-chart/templates/deployment.yaml (hunk 5 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: I know what is the problem, the status.php return 500 because the config-itop.php is missing. 
That file is missing because in kube, when you mount a volume, kubernetes will mount empty directory in `/var/www/html/conf` so the directory with config-itop.php is overwrite by an empty directory.


```
root@itop-itop-chart-79c55995f6-rcvc6:/var/www/html# curl -v localhost/webservices/status.php
*   Trying 127.0.0.1:80...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /webservices/status.php HTTP/1.1
> Host: localhost
> User-Agent: curl/7.74.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 500 Internal Server Error
< Date: Mon, 24 Oct 2022 09:32:30 GMT
< Server: Apache/2.4.54 (Debian)
< X-Powered-By: PHP/7.4.30
< Access-Control-Allow-Origin: *
< Content-Length: 73
< Connection: close
< Content-Type: application/json
< 
* Closing connection 0
{"status":"ERROR","code":100,"message":"config-itop.php is not readable"}
```

Comment (Hipska) [reply helm-chart/templates/deployment.yaml]: Is that also related to the change you made in `deployment.yaml`?

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: No it's diffrent issue. This one will need many changes. 
I'll do tests on my repo and I'll coming back with another pull request for this point. It's ok for you ?

File: helm-chart/templates/deployment.yaml (hunk 6 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: Updated with the `fsgroup`

File: helm-chart/templates/deployment.yaml (hunk 5 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: Roll back, I'll open another PR with that modification

File: helm-chart/templates/deployment.yaml [hunk 10]
```diff
@@ -8,6 +8,8 @@ spec:
   {{- if not .Values.autoscaling.enabled }}
   replicas: {{ .Values.replicaCount }}
   {{- end }}
+  strategy:
+    type: Recreate
```
Comment (Hipska) [inline helm-chart/templates/deployment.yaml:13]: Doesn't the default RollingUpdate work out?

File: helm-chart/templates/deployment.yaml [hunk 11]
```diff
@@ -20,15 +22,13 @@ spec:
...
+        fsGroup: 33
       {{- with .Values.imagePullSecrets }}
       imagePullSecrets:
         {{- toYaml . | nindent 8 }}
       {{- end }}
       serviceAccountName: {{ include "itop-chart.serviceAccountName" . }}
-      securityContext:
-        runAsUser: 1000
-        runAsGroup: 3000
-        fsGroup: 2000
```
Comment (Hipska) [inline helm-chart/templates/deployment.yaml]: Would be nice if this section stays at the same place.

File: helm-chart/templates/deployment.yaml [hunk 12]
```diff
@@ -80,6 +80,7 @@ spec:
               port: http
           resources:
             {{- toYaml .Values.resources | nindent 12 }}
+
       {{- with .Values.nodeSelector }}
```
Comment (Hipska) [inline helm-chart/templates/deployment.yaml]: Empty line not needed?

File: helm-chart/templates/deployment.yaml (hunk 10 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml:13]: Because the new container can't attach the volume if the old container still alive

Comment (Hipska) [reply helm-chart/templates/deployment.yaml:13]: Wait, that sounds like an implementation problem again. How are other projects doing this if they have volumes mounted to the container?

File: helm-chart/templates/deployment.yaml (hunk 11 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: Fixed :)

File: helm-chart/templates/deployment.yaml (hunk 12 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml]: No I removed it :)

**Review by Hipska — APPROVED** [review]

**Review by Hipska — APPROVED** [review]
Thanks, I'll wait for the open conversations to be cleared out until merging.

File: helm-chart/templates/deployment.yaml (hunk 10 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml:13]: with 
```
spec:
  accessModes:
    - ReadWriteOnce
```
Only one container could mount the volume with that config.
If you want rollingUpdate you must have ReadWriteMany PVC

Comment (Hipska) [reply helm-chart/templates/deployment.yaml:13]: Aha okay. What would be a reasonable/working access mode according to you?

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml:13]: We can let this choice to the user, according the storage option available on his infrastructure.
With : 
```
spec:
  accessModes:
    - {{ .Values.pvc.accessModes }} 
 ```
And strategy type according the access modes value

Comment (Hipska) [reply helm-chart/templates/deployment.yaml:13]: Yes, that would be fine for me.

File: helm-chart/values.yaml (hunk 9 above)
Comment (Hipska) [reply helm-chart/values.yaml:61]: @finaldzn Can you elaborate?

File: helm-chart/templates/deployment.yaml (hunk 10 above)
Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml:13]: Pushed, made a if depends on access mode to define stratigy type : 
``` 
 {{-  if eq .Values.pvc.accessMode "ReadWriteOnce" }}
 strategy:
 type: Recreate
 {{- end }} 
```
If not `ReadWriteOne` use default `strategy.type` > `RollingUpdate`

Comment (Hipska) [reply helm-chart/templates/deployment.yaml:13]: Thanks, there only seems to be a problem with pvc.yaml. It looks like the complete file has been recreated? (Different line endings maybe?)

And shouldn't the default `.Values.pvc.accessMode` be `ReadWriteOnce` as it was?

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml:13]: Yes, the file as been created with windows machine and many end line caractere was present, I used dos2unix command. 

The previous access mode was not correct regarding the default strategy type, it was `RWO` and `RollingUpdate`, `RWO` must have `Recreate` type

Comment (Hipska) [reply helm-chart/templates/deployment.yaml:13]: Can you fix the line endings for this file? Another PR can be done to change all file's line endings if needed.

I understand, but the added `if` takes care of that, right?

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml:13]: > Can you fix the line endings for this file? Another PR can be done to change all file's line endings if needed.

Not sur to understand well. 
The pvc.yaml have been fixed, you want me to do same on all other file ? 

> I understand, but the added if takes care of that, right?

Take care of what ? Above end line caratere ?

File: helm-chart/values.yaml (hunk 9 above)
Comment (finaldzn) [reply helm-chart/values.yaml:61]: I was wrong and tought for some reason that this was an other file then values.yaml. My bad sorry :)

File: helm-chart/templates/deployment.yaml (hunk 10 above)
Comment (Hipska) [reply helm-chart/templates/deployment.yaml:13]: 2 different things:

1. I want you to revert the pvc.yaml line endings change.
2. The block below takes care of the accessMode RWO, so IMO that can stay the default value, right?
```yaml
{{-  if eq .Values.pvc.accessMode "ReadWriteOnce" }}
 strategy:
 type: Recreate
 {{- end }} 
```

If you don't agree with RWO as default, then we can see if we just hard-code it to RWM and remove the strategy section alltogether.

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml:13]: Ok I'll revert the pvc. 

Fix an AccessMode is not a good idea, because access mode depend on the storage available on the infrastructure. And ReadWriteMany may not be available. 

In the modification I did, RWM is default (but RWO can also be default as you prefere) is in values.yaml

Comment (Hipska) [reply helm-chart/templates/deployment.yaml:13]: If RWM might not be available, then we definitely should put RWO as default!

Comment (ThomasSamson) [reply helm-chart/templates/deployment.yaml:13]: Right, just wanted to put back your strategy type

Pushed :)

**Review by Hipska — APPROVED** [review]
All looks good! Many thanks.

PS: You are welcome to join the [iTop Slack](https://join.slack.com/t/itopbycombodo/shared_invite/zt-uoat5wuf-C0g_1tqaeI9~EaqHxV7Q1Q)  where you can meet iTop developers and users.
